### PR TITLE
Prune through unnest

### PR DESCRIPTION
**Pull request by JamesRTaylor**
This PR implements two related optimizations that minimize the amount of data read by projecting only the nested columns that are read. For example, if you had the following nested schema for table `T`:

* A
    * a1
        * x
        * y
    * a2
    * a3
* B
    * b1
* C
    * c1

with the following query:
```select x from (select a1 from T cross join unnest(A)) cross join unnest(a1)```

then with this patch, the only Parquet structure that would be projected would be
```array<struct<A:array<struct<a1:array<struct<x:string>>>>>>```

### CODE REVIEW

**Comment by JamesRTaylor** [comment]
@puneetjaiswal - would you have a few cycles to review this patch? I'd like to roll it out to our presto-next cluster so I can see if it helps with performance.

**Comment by JamesRTaylor** [comment]
cc @billonahill @amalakar

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 1]
```diff
@@ -159,6 +159,76 @@ protected TestHiveIntegrationSmokeTest(QueryRunnerSupplier queryRunnerSupplier,
         this.typeTranslator = requireNonNull(typeTranslator, "typeTranslator is null");
     }
 
+    @Test
+    public void testUnnestWithTableScan()
+    {
+        Session admin = Session.builder(getQueryRunner().getDefaultSession())
+                .setIdentity(new Identity("hive", Optional.empty(), ImmutableMap.of("hive", new SelectedRole(SelectedRole.Type.ROLE, Optional.of("admin")))))
+                .build();
+        assertUpdate(admin, "create table nest_test(id int, a array(row(x varchar, y varchar))) WITH (format='PARQUET')");
```
Comment (puneetjaiswal) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:168]: nice!

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneUnusedNestedColumns.java [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+{
+    protected final Metadata metadata;
+
+    public PruneUnusedNestedColumns(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    /**
+     * Extract the nested columns which are being referenced by projectNode
```
Comment (puneetjaiswal) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneUnusedNestedColumns.java:57]: Thanks for adding the example reference, very easy to follow in the code. Wondering if we have documented the design somewhere. Also a good candidate for a blog post :D

**Review by puneetjaiswal — APPROVED** [review]
Very nice!
After rolling this to prod, we can get rid of - https://github.com/lyft/prestoproxy/blob/master/prestoproxy-java/src/main/java/com/lyft/data/proxly/util/QueryBlockingRuleEnforcer.java#L26

**Comment by JamesRTaylor** [comment]
Thanks for the review, @puneetjaiswal. If a query does a `SELECT *` at the top level, this won't help with that, since they user is saying they need *all* of the 1000+ columns returned back to the client. It makes sense to block this, but we can probably get better about if/when it'll be an issue.

**Pull request merged**